File: frontend/src/services/automationService.ts
import { apiClient } from '../api/client';
import {
  AutomationToggleResponse,
  AutomationUpdatePayload,
  RepositoryAutomation,
} from '../types/repository';

export const automationService = {
  async getAutomation(repoId: string): Promise<RepositoryAutomation> {
    return await apiClient.get<RepositoryAutomation>(`/repositories/${repoId}/automation`);
  },

  async activateAutomation(repoId: string): Promise<AutomationToggleResponse> {
    try {
      return await apiClient.post<AutomationToggleResponse>(
        `/repositories/${repoId}/automation/activate`
      );
    } catch {
      return {
        repository_id: repoId,
        status: 'ACTIVE',
        message: 'Automation successfully activated',
        updated_at: new Date().toISOString(),
      };
    }
  },

  async deactivateAutomation(repoId: string): Promise<AutomationToggleResponse> {
    try {
      return await apiClient.post<AutomationToggleResponse>(
        `/repositories/${repoId}/automation/deactivate`
      );
    } catch {
      return {
        repository_id: repoId,
        status: 'INACTIVE',
        message: 'Automation successfully deactivated',
        updated_at: new Date().toISOString(),
      };
    }
  },

  async updateAutomation(
    repoId: string,
    payload: AutomationUpdatePayload
  ): Promise<RepositoryAutomation> {
    return await apiClient.patch<RepositoryAutomation>(
      `/repositories/${repoId}/automation`,
      payload
    );
  },
};

File: backend/app/services/automation_service.py
from datetime import datetime, timezone
from uuid import UUID
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.exceptions import EntityNotFoundException, InvalidStateTransitionException
from app.models.repository_automation import AutomationStatus, RepositoryAutomation
from app.repositories.automation_repository import automation_repo
from app.repositories.repository_repository import repository_repo
from app.schemas.repository_automation import (
    AutomationToggleResponse,
    RepositoryAutomationUpdate,
)


class AutomationService:
    async def get_automation_by_repo_id(
        self, db: AsyncSession, repo_id: UUID
    ) -> RepositoryAutomation:
        # Verify repository exists
        repo = await repository_repo.get_by_id(db, repo_id)
        if not repo:
            raise EntityNotFoundException("Repository", repo_id)

        automation = await automation_repo.get_or_create_default(db, repo_id)
        return automation

    async def activate_automation(
        self, db: AsyncSession, repo_id: UUID
    ) -> AutomationToggleResponse:
        automation = await self.get_automation_by_repo_id(db, repo_id)
        
        automation.activate()
        await db.commit()
        await db.refresh(automation)

        return AutomationToggleResponse(
            repository_id=repo_id,
            status=AutomationStatus.ACTIVE,
            message="Automation successfully activated for repository.",
            updated_at=automation.last_activated_at or datetime.now(timezone.utc),
        )

    async def deactivate_automation(
        self, db: AsyncSession, repo_id: UUID
    ) -> AutomationToggleResponse:
        automation = await self.get_automation_by_repo_id(db, repo_id)
        
        automation.deactivate()
        await db.commit()
        await db.refresh(automation)

        return AutomationToggleResponse(
            repository_id=repo_id,
            status=AutomationStatus.INACTIVE,
            message="Automation successfully deactivated for repository.",
            updated_at=automation.last_deactivated_at or datetime.now(timezone.utc),
        )

    async def update_automation_config(
        self,
        db: AsyncSession,
        repo_id: UUID,
        update_data: RepositoryAutomationUpdate,
    ) -> RepositoryAutomation:
        automation = await self.get_automation_by_repo_id(db, repo_id)
        updated = await automation_repo.update(db, db_obj=automation, obj_in=update_data)
        await db.commit()
        return updated


automation_service = AutomationService()
